internal/generate/crd/crd.go

```diff
@@ -0,0 +1,332 @@
+// for Go and non-Go projects.
+type crdGenerator struct {
+	genutil.Config
+	// isOperatorGo is true when the operator is written in Go.
+	isOperatorGo bool
+	// resource contains API information used to configure single-CRD generation.
+	// This is only required when isOperatorGo is false.
+	resource scaffold.Resource
+}
+
+// NewCRDGo returns a CRD generator configured to generate CustomResourceDefintion
+// manifests from Go API files.
+func NewCRDGo(cfg genutil.Config) genutil.Generator {
+	g := crdGenerator{
+		Config:       cfg,
+		isOperatorGo: true,
+	}
+	if g.InputDir == "" {
+		g.InputDir = scaffold.ApisDir
+	}
+	if g.OutputDir == "" {
+		g.OutputDir = scaffold.CRDsDir
+	}
+	return g
+}
+
+// NewCRDGo returns a CRD generator configured to generate a
+// CustomResourceDefintion manifest from scratch using data in resource.
+func NewCRDNonGo(cfg genutil.Config, resource scaffold.Resource) genutil.Generator {
+	g := crdGenerator{
+		Config:       cfg,
+		resource:     resource,
+		isOperatorGo: false,
+	}
+	if g.InputDir == "" {
+		g.InputDir = scaffold.CRDsDir
+	}
+	if g.OutputDir == "" {
+		g.OutputDir = scaffold.CRDsDir
+	}
+	return g
+}
+
+func (g crdGenerator) validate() error {
+	if g.InputDir == "" {
+		return errors.New("input dir cannot be empty")
+	}
+	if g.InputDir == "" {
+		return errors.New("output dir cannot be empty")
+	}
+	if !g.isOperatorGo {
+		if err := g.resource.Validate(); err != nil {
+			return errors.Wrap(err, "resource is invalid:")
+		}
+	}
+	return nil
+}
+
+// Generate generates CRD manifests and writes them to g.OutputDir.
+func (g crdGenerator) Generate() (err error) {
+	if err = g.validate(); err != nil {
+		return errors.Wrap(err, "validation error")
+	}
+	var fileMap map[string][]byte
+	if g.isOperatorGo {
+		fileMap, err = g.generateGo()
+	} else {
+		fileMap, err = g.generateNonGo()
+	}
+	if err != nil {
+		return errors.Wrap(err, "error generating CRD manifests")
+	}
+	if err = os.MkdirAll(g.OutputDir, fileutil.DefaultDirFileMode); err != nil {
+		return errors.Wrapf(err, "error mkdir %s", g.OutputDir)
+	}
+	for fileName, b := range fileMap {
+		path := filepath.Join(g.OutputDir, fileName)
+		if err := ioutil.WriteFile(path, b, fileutil.DefaultFileMode); err != nil {
+			return errors.Wrap(err, "error writing CRD manifests")
+		}
+	}
+	return nil
+}
+
+func getFileNameForResource(r scaffold.Resource) string {
+	return fmt.Sprintf("%s_%s_crd.yaml", r.FullGroup, r.Resource)
+}
+
+// generateNonGo generates CRDs for Go projects using Go API files.
+func (g crdGenerator) generateGo() (map[string][]byte, error) {
+	fileMap := map[string][]byte{}
+	// Generate files in the generator's cache so we can modify the file name
+	// and annotations.
+	defName := "output:crd:cache"
+	cacheOutputDir := string(filepath.Separator) + filepath.Clean(g.OutputDir)
+	rawOpts := []string{
+		"crd",
+		fmt.Sprintf("paths=%s/...", fileutil.DotPath(g.InputDir)),
```

Comment:
Do we need to use `filepath.Separator` in `%s/...` instead of the forward slash? Or does Go always expect the forward slash for this style of path expression?

Reply:
`/` is always used for package paths.